### PR TITLE
MAID-2488 feat/webFetch: impl multipart range requests

### PR DESCRIPTION
**Pull request by hunterlester**
See commit, https://github.com/joshuef/peruse/commit/dff8f2151570be8cefce19bfb0f93baf31c1870c, of this PR and also https://github.com/maidsafe/safe_app_nodejs/pull/223

When making a multipart range request with `webFetch` Peruse will receive a response as follows:
```{
  "headers": {
       "Content-Type": "multipart/byteranges",
       "Content-Length": 200
   }.
   parts: [
       {
           "headers": {
                "Content-Type: "video/mp4,
                "Content-Range":  bytes 4-15/200
            },
            body: <Buffer>
       },
        {
           "headers": {
                "Content-Type: "video/mp4,
                "Content-Range":  bytes 40-99/200
            },
            body: <Buffer>
       },
        {
           "headers": {
                "Content-Type: "video/mp4,
                "Content-Range":  bytes 150-175/200
            },
            body: <Buffer>
       },
   ]
}
```

### CODE REVIEW

File: app/extensions/safe/utils/safeHelpers.js [hunk 1]
```diff
@@ -27,3 +27,14 @@ export const urlIsAllowed = ( testUrl ) =>
 
     return false;
 };
+
+export const generateBoundaryStr = () => {
```
Comment (joshuef) [inline app/extensions/safe/utils/safeHelpers.js]: Lets add some tests for this func

File: app/extensions/safe/server-routes/safe.js [hunk 2]
```diff
@@ -28,34 +31,53 @@ const safeRoute = ( store ) => ( {
...
-                {
-                    // should have a start !== 0
-                    isRangeReq = true;
-                }
-
-                if ( start === 0 && end )
-                {
-                    isRangeReq = true;
-                }
+                      .split( ',' )
```
Comment (joshuef) [inline app/extensions/safe/server-routes/safe.js]: @hunterlester it's looking like this function is getting loooong.

Can we start splitting this out (maybe into another file if it makes sense?) To keep the overall server handling clean.

We should then be able to more easily test that it's working correctly too 👍

File: app/extensions/safe/server-routes/safe.js [hunk 3]
```diff
@@ -75,19 +97,38 @@ const safeRoute = ( store ) => ( {
             }
             store.dispatch( setWebFetchStatus( { fetching: false, options: '' } ) );
 
-            if ( isRangeReq )
+            if ( isRangeReq && multipartReq )
+            {
+		const boundaryStr = generateBoundaryStr();
```
Comment (joshuef) [inline app/extensions/safe/server-routes/safe.js]: I think this is ripe for being its own func too

**Review by joshuef — CHANGES_REQUESTED** [review]

**Comment by joshuef** [comment]
Actually, @hunterlester, is this up to date with the latest changes made before the merge? (for using `parts` in the response?)

**Comment by hunterlester** [comment]
I know we're keeping inline with HTTP standards by returning a response string for multipart requests but, Marone, it's really an unfriendly format and in some cases may be unusable.
For instance, if a request for multiple parts of a stream is made, a response stream will be returned with the those byte segments transformed to utf-8 characters, which may not make sense for the original raw byte encoding.

Here's a multipart response string example for a `Buffer` created from the following string: `Lorem ipsum dolor sit amet, consectetur adipiscing elit`
```
HTTP/1.1 206 Partial Content
Content-Type: multipart/byteranges; boundary=axk2jtnn13t5i
Content-Length:367
--axk2jtnn13t5i
Content-Type:text/plain
Content-Range: bytes 3-8/55
em ips
--axk2jtnn13t5i
Content-Type:text/plain
Content-Range: bytes 11-13/55
 do
--axk2jtnn13t5i
Content-Type:text/plain
Content-Range: bytes 17-18/55
 s
--axk2jtnn13t5i--
```

This format happens to be kind of okay since the contents of the `Buffer` is composed of utf-8 encoded bytes but why add the extra step and make the consumer have to parse the response when `webFetch` is already returning a convenient format:
```
{ headers:
   { 'Content-Type': 'multipart/byteranges',
     'Content-Length': 367 },
  parts:
   [ { body: <Buffer 65 6d 20 69 70 73>, headers: [Object] },
     { body: <Buffer 20 64 6f>, headers: [Object] },
     { body: <Buffer 20 73>, headers: [Object] } ] }
```

I would think that the primary use case for multipart requests would be for streams though, which makes the response string make even less sense.

cc: @joshuef @bochaco

**Comment by bochaco** [comment]
@hunterlester , the consumer of such a response is not any app but an app trying to fetch content using the http protocol and headers. If we change the protocol the video player of HTML5 wouldn't work and we would have to provide our own implementation for it in the browser. Moreover, any content we return is http complaint so the browser's window can render it correctly.

**Review by joshuef — APPROVED** [review]

**Pull request merged**